### PR TITLE
implement ERC721Enumerable

### PR DESCRIPTION
**Pull request by NicolasMahe**
closes https://github.com/QuiverCommunity/quiver-contracts/issues/19

This PR implements `ERC721Enumerable`.

In order to do this I had to rename `totalSupply` to `maxSupply` and update the logic to use the new available functions from `ERC721Enumerable`.

- Now `totalSupply` (implemented by `ERC721Enumerable`) is the actual number of existing NFT (not the max like before).
- `maxSupply` is the maximum number of NFT that can be created.

Like the other PR, I did not test. @junkai121 @gopherquiv @icrabbiter please review and test strongly

### CODE REVIEW

File: contracts/QNFT/QNFT.sol [hunk 1]
```diff
@@ -22,7 +22,7 @@ import "../interface/IQAirdrop.sol";
 contract QNFT is
     IQNFT,
     ContextUpgradeable,
-    ERC721Upgradeable,
+    ERC721EnumerableUpgradeable,
```
Comment (NicolasMahe) [inline contracts/QNFT/QNFT.sol:25]: `ERC721EnumerableUpgradeable` is already an `ERC721Upgradeable`

File: contracts/QNFT/QNFT.sol [hunk 2]
```diff
@@ -33,7 +33,7 @@ contract QNFT is
     event RemoveFreeAllocation(address indexed user, uint256 amount);
     event DepositQstk(address indexed owner, uint256 amount);
     event WithdrawQstk(address indexed owner, uint256 amount);
-    event SetTotalSupply(address indexed owner, uint256 totalSupply);
+    event SetMaxSupply(uint256 maxSupply);
```
Comment (NicolasMahe) [inline contracts/QNFT/QNFT.sol:36]: I removed the `owner` from this event as it will always be the `manager`

File: contracts/QNFT/QNFT.sol [hunk 3]
```diff
@@ -174,19 +173,12 @@ contract QNFT is
 
     // NFT
 
-    /**
-     * @dev returns minted nft count
-     */
-    function circulatingSupply() public view returns (uint256) {
-        return nftCount;
-    }
-
```
Comment (NicolasMahe) [inline contracts/QNFT/QNFT.sol:183]: this function is implemented by `ERC721EnumerableUpgradeable` as `totalSupply()`

File: contracts/QNFT/QNFT.sol [hunk 4]
```diff
@@ -523,8 +515,8 @@ contract QNFT is
             "QNFT: nft already exists"
         );
 
-        nftCount++;
-        nftData[nftCount] = NFTData(
+        uint256 newId = totalSupply() + 1;
```
Comment (NicolasMahe) [inline contracts/QNFT/QNFT.sol:518]: This feels weird for me but make sense as you want to start the token's id at 1 instead of the classic 0.

Comment (gopherquiv) [reply contracts/QNFT/QNFT.sol:518]: yes we want to start from 1

File: contracts/QNFT/QNFT.sol (hunk 3 above)
Comment (gopherquiv) [reply contracts/QNFT/QNFT.sol:183]: I see

File: contracts/QNFT/QNFT.sol (hunk 2 above)
Comment (gopherquiv) [reply contracts/QNFT/QNFT.sol:36]: Ok

File: contracts/QNFT/QNFT.sol (hunk 1 above)
Comment (gopherquiv) [reply contracts/QNFT/QNFT.sol:25]: Ok

**Comment by gopherquiv** [comment]
@icrabbiter  can you check finalize this PR

**Comment by icrabbiter** [comment]
Looks good to me!

File: contracts/QNFT/QNFT.sol (hunk 4 above)
Comment (gopherquiv) [reply contracts/QNFT/QNFT.sol:518]: @icrabbiter  do you think 0 is better?

**Review by icrabbiter — APPROVED** [review]

**Pull request merged**